diff --git a/lib/color.c b/lib/color.c
index f62389d5..1f97c6c9 100644
--- a/lib/color.c
+++ b/lib/color.c
@@ -32,7 +32,7 @@
 #endif
 
 #include <argp.h>
-#include <error.h>
+#include <compat.h>
 #include <libintl.h>
 #include <stdlib.h>
 #include <string.h>
diff --git a/lib/xmalloc.c b/lib/xmalloc.c
index 0cde384f..5de9c2eb 100644
--- a/lib/xmalloc.c
+++ b/lib/xmalloc.c
@@ -30,7 +30,7 @@
 # include <config.h>
 #endif
 
-#include <error.h>
+#include <compat.h>
 #include <libintl.h>
 #include <stddef.h>
 #include <stdlib.h>
diff --git a/libasm/asm_end.c b/libasm/asm_end.c
index ced24f50..fad6d046 100644
--- a/libasm/asm_end.c
+++ b/libasm/asm_end.c
@@ -32,7 +32,7 @@
 #endif
 
 #include <assert.h>
-#include <error.h>
+#include <compat.h>
 #include <libintl.h>
 #include <stdio.h>
 #include <stdlib.h>
diff --git a/libasm/asm_newscn.c b/libasm/asm_newscn.c
index ddbb25df..eee1f7d2 100644
--- a/libasm/asm_newscn.c
+++ b/libasm/asm_newscn.c
@@ -32,7 +32,7 @@
 #endif
 
 #include <assert.h>
-#include <error.h>
+#include <compat.h>
 #include <libintl.h>
 #include <stdlib.h>
 #include <string.h>
diff --git a/libcpu/i386_gendis.c b/libcpu/i386_gendis.c
index aae5eae6..632b73a3 100644
--- a/libcpu/i386_gendis.c
+++ b/libcpu/i386_gendis.c
@@ -31,7 +31,7 @@
 # include <config.h>
 #endif
 
-#include <error.h>
+#include <compat.h>
 #include <errno.h>
 #include <stdio.h>
 #include <stdlib.h>
diff --git a/libdw/libdw_alloc.c b/libdw/libdw_alloc.c
index d6af23a2..efe2d5d3 100644
--- a/libdw/libdw_alloc.c
+++ b/libdw/libdw_alloc.c
@@ -31,7 +31,7 @@
 # include <config.h>
 #endif
 
-#include <error.h>
+#include <compat.h>
 #include <errno.h>
 #include <stdlib.h>
 #include "libdwP.h"
diff --git a/libebl/eblopenbackend.c b/libebl/eblopenbackend.c
index 1f814776..cc81f472 100644
--- a/libebl/eblopenbackend.c
+++ b/libebl/eblopenbackend.c
@@ -32,7 +32,7 @@
 
 #include <assert.h>
 #include <dlfcn.h>
-#include <error.h>
+#include <compat.h>
 #include <libelfP.h>
 #include <dwarf.h>
 #include <stdlib.h>
diff --git a/src/addr2line.c b/src/addr2line.c
index ba414a74..255a3a6d 100644
--- a/src/addr2line.c
+++ b/src/addr2line.c
@@ -23,7 +23,7 @@
 #include <argp.h>
 #include <assert.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <inttypes.h>
 #include <libdwfl.h>
diff --git a/src/ar.c b/src/ar.c
index 818115bd..feff63ff 100644
--- a/src/ar.c
+++ b/src/ar.c
@@ -22,7 +22,7 @@
 
 #include <argp.h>
 #include <assert.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <gelf.h>
 #include <libintl.h>
diff --git a/src/arlib.c b/src/arlib.c
index e0839aab..a622fce4 100644
--- a/src/arlib.c
+++ b/src/arlib.c
@@ -21,7 +21,7 @@
 #endif
 
 #include <assert.h>
-#include <error.h>
+#include <compat.h>
 #include <gelf.h>
 #include <inttypes.h>
 #include <libintl.h>
diff --git a/src/arlib2.c b/src/arlib2.c
index 553fc57b..79027f2b 100644
--- a/src/arlib2.c
+++ b/src/arlib2.c
@@ -20,7 +20,7 @@
 # include <config.h>
 #endif
 
-#include <error.h>
+#include <compat.h>
 #include <libintl.h>
 #include <limits.h>
 #include <string.h>
diff --git a/compat.h b/compat.h
new file mode 100644
index 00000000..f449febe
--- /dev/null
+++ b/compat.h
@@ -0,0 +1,39 @@
+#ifndef _COMPAT_H
+#define _COMPAT_H
+
+#ifdef HAVE_ERROR_H
+# include <error.h>
+#else
+# include <stdio.h>
+# include <stdarg.h>
+# include <stdlib.h>
+# include <string.h>
+static void error_at_line(int status, int errnum, const char *filename,
+                          unsigned int linenum, const char *format, ...)
+{
+	va_list ap;
+
+	fflush(stdout);
+
+	if (filename != NULL)
+		fprintf(stderr, "%s:%u: ", filename, linenum);
+
+	va_start(ap, format);
+	vfprintf(stderr, format, ap);
+	va_end(ap);
+
+	if (errnum != 0)
+		fprintf(stderr, ": %s", strerror(errnum));
+
+	fprintf(stderr, "\n");
+
+	if (status != 0)
+		exit(status);
+}
+
+#define error(status, errnum, format...) \
+	error_at_line(status, errnum, NULL, 0, format)
+
+#endif /* HAVE_ERROR_H */
+
+#endif /* _COMPAT_H */
diff --git a/src/elfcmp.c b/src/elfcmp.c
index 50464207..2dccbcfd 100644
--- a/src/elfcmp.c
+++ b/src/elfcmp.c
@@ -23,7 +23,7 @@
 #include <argp.h>
 #include <assert.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <locale.h>
 #include <libintl.h>
diff --git a/src/elfcompress.c b/src/elfcompress.c
index 8e0d5c55..0ac92104 100644
--- a/src/elfcompress.c
+++ b/src/elfcompress.c
@@ -18,7 +18,7 @@
 #include <config.h>
 #include <assert.h>
 #include <argp.h>
-#include <error.h>
+#include <compat.h>
 #include <stdbool.h>
 #include <stdlib.h>
 #include <inttypes.h>
diff --git a/src/elflint.c b/src/elflint.c
index 51e53c23..5aeeabf6 100644
--- a/src/elflint.c
+++ b/src/elflint.c
@@ -24,7 +24,7 @@
 #include <assert.h>
 #include <byteswap.h>
 #include <endian.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <gelf.h>
 #include <inttypes.h>
diff --git a/src/findtextrel.c b/src/findtextrel.c
index 8f1e239a..96db0085 100644
--- a/src/findtextrel.c
+++ b/src/findtextrel.c
@@ -23,7 +23,7 @@
 #include <argp.h>
 #include <assert.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <gelf.h>
 #include <libdw.h>
diff --git a/src/nm.c b/src/nm.c
index 969c6d35..98c1a807 100644
--- a/src/nm.c
+++ b/src/nm.c
@@ -26,7 +26,7 @@
 #include <ctype.h>
 #include <dwarf.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <gelf.h>
 #include <inttypes.h>
diff --git a/src/objdump.c b/src/objdump.c
index 860cfac6..5de57d6f 100644
--- a/src/objdump.c
+++ b/src/objdump.c
@@ -21,7 +21,7 @@
 #endif
 
 #include <argp.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <inttypes.h>
 #include <libintl.h>
diff --git a/src/ranlib.c b/src/ranlib.c
index cc0ee233..f99f57cf 100644
--- a/src/ranlib.c
+++ b/src/ranlib.c
@@ -24,7 +24,7 @@
 #include <argp.h>
 #include <assert.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <gelf.h>
 #include <libintl.h>
diff --git a/src/readelf.c b/src/readelf.c
index 6c49d305..fb47da0c 100644
--- a/src/readelf.c
+++ b/src/readelf.c
@@ -25,7 +25,7 @@
 #include <ctype.h>
 #include <dwarf.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <gelf.h>
 #include <inttypes.h>
diff --git a/src/size.c b/src/size.c
index ad8dbcbb..b593f7e8 100644
--- a/src/size.c
+++ b/src/size.c
@@ -21,7 +21,7 @@
 #endif
 
 #include <argp.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <gelf.h>
 #include <inttypes.h>
diff --git a/src/stack.c b/src/stack.c
index 52ae3a86..49f597d9 100644
--- a/src/stack.c
+++ b/src/stack.c
@@ -18,7 +18,7 @@
 #include <config.h>
 #include <assert.h>
 #include <argp.h>
-#include <error.h>
+#include <compat.h>
 #include <stdlib.h>
 #include <inttypes.h>
 #include <stdio.h>
diff --git a/src/strings.c b/src/strings.c
index d214356c..ba863a2b 100644
--- a/src/strings.c
+++ b/src/strings.c
@@ -25,7 +25,7 @@
 #include <ctype.h>
 #include <endian.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <gelf.h>
 #include <inttypes.h>
diff --git a/src/strip.c b/src/strip.c
index 773ed548..9928f494 100644
--- a/src/strip.c
+++ b/src/strip.c
@@ -24,7 +24,7 @@
 #include <assert.h>
 #include <byteswap.h>
 #include <endian.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <fnmatch.h>
 #include <gelf.h>
diff --git a/src/unstrip.c b/src/unstrip.c
index f368e696..4c602b23 100644
--- a/src/unstrip.c
+++ b/src/unstrip.c
@@ -31,7 +31,7 @@
 #include <argp.h>
 #include <assert.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <fnmatch.h>
 #include <libintl.h>
diff --git a/tests/addrscopes.c b/tests/addrscopes.c
index 791569f5..ab2c1290 100644
--- a/tests/addrscopes.c
+++ b/tests/addrscopes.c
@@ -25,7 +25,7 @@
 #include <stdio_ext.h>
 #include <locale.h>
 #include <stdlib.h>
-#include <error.h>
+#include <compat.h>
 #include <string.h>
 
 
diff --git a/tests/allregs.c b/tests/allregs.c
index 286f7e3c..2607e406 100644
--- a/tests/allregs.c
+++ b/tests/allregs.c
@@ -21,7 +21,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <error.h>
+#include <compat.h>
 #include <locale.h>
 #include <argp.h>
 #include <assert.h>
diff --git a/tests/backtrace-data.c b/tests/backtrace-data.c
index a387d8ff..3df87528 100644
--- a/tests/backtrace-data.c
+++ b/tests/backtrace-data.c
@@ -27,7 +27,7 @@
 #include <dirent.h>
 #include <stdlib.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <unistd.h>
 #include <dwarf.h>
 #if defined(__x86_64__) && defined(__linux__)
diff --git a/tests/backtrace-dwarf.c b/tests/backtrace-dwarf.c
index 2dc8a9a2..6ee3a99b 100644
--- a/tests/backtrace-dwarf.c
+++ b/tests/backtrace-dwarf.c
@@ -22,7 +22,7 @@
 #include <stdio_ext.h>
 #include <locale.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <unistd.h>
 #include <sys/types.h>
 #include <sys/wait.h>
diff --git a/tests/backtrace.c b/tests/backtrace.c
index 21abe8af..1385d60f 100644
--- a/tests/backtrace.c
+++ b/tests/backtrace.c
@@ -24,7 +24,7 @@
 #include <dirent.h>
 #include <stdlib.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <unistd.h>
 #include <dwarf.h>
 #ifdef __linux__
diff --git a/tests/buildid.c b/tests/buildid.c
index 87c18773..2c2bdd70 100644
--- a/tests/buildid.c
+++ b/tests/buildid.c
@@ -23,7 +23,7 @@
 #include ELFUTILS_HEADER(elf)
 #include ELFUTILS_HEADER(dwelf)
 #include <stdio.h>
-#include <error.h>
+#include <compat.h>
 #include <string.h>
 #include <stdlib.h>
 #include <sys/types.h>
diff --git a/tests/debugaltlink.c b/tests/debugaltlink.c
index 6d97d500..c07cae33 100644
--- a/tests/debugaltlink.c
+++ b/tests/debugaltlink.c
@@ -23,7 +23,7 @@
 #include ELFUTILS_HEADER(dw)
 #include ELFUTILS_HEADER(dwelf)
 #include <stdio.h>
-#include <error.h>
+#include <compat.h>
 #include <string.h>
 #include <stdlib.h>
 #include <sys/types.h>
diff --git a/tests/debuglink.c b/tests/debuglink.c
index 935d1029..204a5d1e 100644
--- a/tests/debuglink.c
+++ b/tests/debuglink.c
@@ -21,7 +21,7 @@
 #include <errno.h>
 #include ELFUTILS_HEADER(dwelf)
 #include <stdio.h>
-#include <error.h>
+#include <compat.h>
 #include <string.h>
 #include <stdlib.h>
 #include <sys/types.h>
diff --git a/tests/deleted.c b/tests/deleted.c
index 6be35bc2..2078e745 100644
--- a/tests/deleted.c
+++ b/tests/deleted.c
@@ -21,7 +21,7 @@
 #include <unistd.h>
 #include <assert.h>
 #include <stdio.h>
-#include <error.h>
+#include <compat.h>
 #include <errno.h>
 #ifdef __linux__
 #include <sys/prctl.h>
diff --git a/tests/dwfl-addr-sect.c b/tests/dwfl-addr-sect.c
index 21e470a3..696c3cbb 100644
--- a/tests/dwfl-addr-sect.c
+++ b/tests/dwfl-addr-sect.c
@@ -23,7 +23,7 @@
 #include <stdio_ext.h>
 #include <stdlib.h>
 #include <string.h>
-#include <error.h>
+#include <compat.h>
 #include <locale.h>
 #include <argp.h>
 #include ELFUTILS_HEADER(dwfl)
diff --git a/tests/dwfl-bug-addr-overflow.c b/tests/dwfl-bug-addr-overflow.c
index aa8030e1..7bb639c4 100644
--- a/tests/dwfl-bug-addr-overflow.c
+++ b/tests/dwfl-bug-addr-overflow.c
@@ -20,7 +20,7 @@
 #include <inttypes.h>
 #include <stdio.h>
 #include <stdio_ext.h>
-#include <error.h>
+#include <compat.h>
 #include <locale.h>
 #include ELFUTILS_HEADER(dwfl)
 
diff --git a/tests/dwfl-bug-fd-leak.c b/tests/dwfl-bug-fd-leak.c
index 689cdd79..d81ff521 100644
--- a/tests/dwfl-bug-fd-leak.c
+++ b/tests/dwfl-bug-fd-leak.c
@@ -24,7 +24,7 @@
 #include <dirent.h>
 #include <stdlib.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <unistd.h>
 #include <dwarf.h>
 
diff --git a/tests/dwfl-bug-getmodules.c b/tests/dwfl-bug-getmodules.c
index 1ee989f8..4015a0af 100644
--- a/tests/dwfl-bug-getmodules.c
+++ b/tests/dwfl-bug-getmodules.c
@@ -18,7 +18,7 @@
 #include <config.h>
 #include ELFUTILS_HEADER(dwfl)
 
-#include <error.h>
+#include <compat.h>
 
 static const Dwfl_Callbacks callbacks =
   {
diff --git a/tests/dwfl-proc-attach.c b/tests/dwfl-proc-attach.c
index e7bb2010..5701f58d 100644
--- a/tests/dwfl-proc-attach.c
+++ b/tests/dwfl-proc-attach.c
@@ -20,7 +20,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <unistd.h>
 #ifdef __linux__
 #include <sys/types.h>
diff --git a/tests/dwfl-report-elf-align.c b/tests/dwfl-report-elf-align.c
index a4e97d3c..ffd146ba 100644
--- a/tests/dwfl-report-elf-align.c
+++ b/tests/dwfl-report-elf-align.c
@@ -20,7 +20,7 @@
 #include <inttypes.h>
 #include <stdio.h>
 #include <stdio_ext.h>
-#include <error.h>
+#include <compat.h>
 #include <locale.h>
 #include <string.h>
 #include <stdlib.h>
diff --git a/tests/dwfllines.c b/tests/dwfllines.c
index 90379dd2..df6c9d04 100644
--- a/tests/dwfllines.c
+++ b/tests/dwfllines.c
@@ -27,7 +27,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <error.h>
+#include <compat.h>
 
 int
 main (int argc, char *argv[])
diff --git a/tests/dwflmodtest.c b/tests/dwflmodtest.c
index 0027f96b..469b9a94 100644
--- a/tests/dwflmodtest.c
+++ b/tests/dwflmodtest.c
@@ -23,7 +23,7 @@
 #include <stdio_ext.h>
 #include <stdlib.h>
 #include <string.h>
-#include <error.h>
+#include <compat.h>
 #include <locale.h>
 #include <argp.h>
 #include ELFUTILS_HEADER(dwfl)
diff --git a/tests/dwflsyms.c b/tests/dwflsyms.c
index 49ac3346..a3dc6b6a 100644
--- a/tests/dwflsyms.c
+++ b/tests/dwflsyms.c
@@ -25,7 +25,7 @@
 #include <stdio.h>
 #include <stdio_ext.h>
 #include <stdlib.h>
-#include <error.h>
+#include <compat.h>
 #include <string.h>
 
 static const char *
diff --git a/tests/early-offscn.c b/tests/early-offscn.c
index 924cb9ef..883aca81 100644
--- a/tests/early-offscn.c
+++ b/tests/early-offscn.c
@@ -19,7 +19,7 @@
 #endif
 
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <gelf.h>
 #include <stdio.h>
diff --git a/tests/ecp.c b/tests/ecp.c
index 38a6859e..87ab14a3 100644
--- a/tests/ecp.c
+++ b/tests/ecp.c
@@ -20,7 +20,7 @@
 #endif
 
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <gelf.h>
 #include <stdlib.h>
diff --git a/tests/elfstrmerge.c b/tests/elfstrmerge.c
index 6924d0e9..81dd089c 100644
--- a/tests/elfstrmerge.c
+++ b/tests/elfstrmerge.c
@@ -24,7 +24,7 @@
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <fcntl.h>
-#include <error.h>
+#include <compat.h>
 #include <stdio.h>
 #include <inttypes.h>
 #include <unistd.h>
diff --git a/tests/find-prologues.c b/tests/find-prologues.c
index ba8ae371..0f1f5443 100644
--- a/tests/find-prologues.c
+++ b/tests/find-prologues.c
@@ -25,7 +25,7 @@
 #include <stdio_ext.h>
 #include <locale.h>
 #include <stdlib.h>
-#include <error.h>
+#include <compat.h>
 #include <string.h>
 #include <fnmatch.h>
 
diff --git a/tests/funcretval.c b/tests/funcretval.c
index 8d19d117..73b9532c 100644
--- a/tests/funcretval.c
+++ b/tests/funcretval.c
@@ -25,7 +25,7 @@
 #include <stdio_ext.h>
 #include <locale.h>
 #include <stdlib.h>
-#include <error.h>
+#include <compat.h>
 #include <string.h>
 #include <fnmatch.h>
 
diff --git a/tests/funcscopes.c b/tests/funcscopes.c
index 9c901858..b2b5f17b 100644
--- a/tests/funcscopes.c
+++ b/tests/funcscopes.c
@@ -25,7 +25,7 @@
 #include <stdio_ext.h>
 #include <locale.h>
 #include <stdlib.h>
-#include <error.h>
+#include <compat.h>
 #include <string.h>
 #include <fnmatch.h>
 
diff --git a/tests/getsrc_die.c b/tests/getsrc_die.c
index 055aede0..d093fc04 100644
--- a/tests/getsrc_die.c
+++ b/tests/getsrc_die.c
@@ -19,7 +19,7 @@
 #endif
 
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <inttypes.h>
 #include <libelf.h>
diff --git a/tests/line2addr.c b/tests/line2addr.c
index e0d65d3d..953aca9a 100644
--- a/tests/line2addr.c
+++ b/tests/line2addr.c
@@ -26,7 +26,7 @@
 #include <locale.h>
 #include <stdlib.h>
 #include <string.h>
-#include <error.h>
+#include <compat.h>
 
 
 static void
diff --git a/tests/low_high_pc.c b/tests/low_high_pc.c
index d0f43023..ef0f8aa8 100644
--- a/tests/low_high_pc.c
+++ b/tests/low_high_pc.c
@@ -25,7 +25,7 @@
 #include <stdio_ext.h>
 #include <locale.h>
 #include <stdlib.h>
-#include <error.h>
+#include <compat.h>
 #include <string.h>
 #include <fnmatch.h>
 
diff --git a/tests/rdwrmmap.c b/tests/rdwrmmap.c
index 6f027dfe..0b9b940e 100644
--- a/tests/rdwrmmap.c
+++ b/tests/rdwrmmap.c
@@ -19,7 +19,7 @@
 #endif
 
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <stdio.h>
 #include <fcntl.h>
 #include <unistd.h>
diff --git a/tests/saridx.c b/tests/saridx.c
index 8a450d82..47de6ef4 100644
--- a/tests/saridx.c
+++ b/tests/saridx.c
@@ -17,7 +17,7 @@
 
 #include <config.h>
 
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <gelf.h>
 #include <stdio.h>
diff --git a/tests/sectiondump.c b/tests/sectiondump.c
index 3033fedc..0e3c510c 100644
--- a/tests/sectiondump.c
+++ b/tests/sectiondump.c
@@ -18,7 +18,7 @@
 #include <config.h>
 
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <fcntl.h>
 #include <gelf.h>
 #include <inttypes.h>
diff --git a/tests/varlocs.c b/tests/varlocs.c
index 0e432296..b22cf821 100644
--- a/tests/varlocs.c
+++ b/tests/varlocs.c
@@ -25,7 +25,7 @@
 #include <dwarf.h>
 #include <stdio.h>
 #include <stdlib.h>
-#include <error.h>
+#include <compat.h>
 #include <string.h>
 #include <sys/types.h>
 #include <sys/stat.h>
diff --git a/tests/vdsosyms.c b/tests/vdsosyms.c
index b876c10b..66461b35 100644
--- a/tests/vdsosyms.c
+++ b/tests/vdsosyms.c
@@ -18,7 +18,7 @@
 #include <config.h>
 #include <assert.h>
 #include <errno.h>
-#include <error.h>
+#include <compat.h>
 #include <inttypes.h>
 #include <stdio.h>
 #include <string.h>
